File: 3201-find-the-maximum-length-of-valid-subsequence-i/3201-find-the-maximum-length-of-valid-subsequence-i.cpp
class Solution {
public:
    int maximumLength(vector<int>& nums) {
        int olen = 0;
        int elen = 0;
        int altlen1 = 0;
        int altlen2 = 0;
        bool a1 = true;
        bool a2 = false;
        for (auto x: nums){
            if (x%2 == 1){
                olen++;
                if (a1){
                    altlen1++;
                    a1 = !a1;
                }
                if (a2){
                    altlen2++;
                    a2 = !a2;
                }
            }
            else{
                elen++;
                if (!a1){
                    altlen1++;
                    a1 = !a1;
                }
                if (!a2){
                    altlen2++;
                    a2 = !a2;
                }
            }
        }
        return (max(max(olen,elen),max(altlen1,altlen2)));
    }
};

File: 3201-find-the-maximum-length-of-valid-subsequence-i/3201-find-the-maximum-length-of-valid-subsequence-i.py
class Solution:
    def maximumLength(self, nums: List[int]) -> int:
        def helper(first, second):
            if len(nums) <= 2:
                return len(nums)
            search = first
            res = 0
            for i in range(len(nums)):
                if nums[i] % 2 == search:
                    res += 1
                    search = first if search == second else second
            return res

        return max(helper(0, 0), helper(0, 1), helper(1, 0), helper(1, 1))